### PR TITLE
Enhancement shim datapath

### PR DESCRIPTION
**Pull request by stolet**
Adding a batch of optimizations for the shim layer
- Use static arrays instead of custom hashtable and hashset
- Hint to the compiler likely and unlikely events in branch conditions
- Add missing O3 flag to the shim makefile
- Use demi_wait_any instead of regular demi_wait to reduce the number of demi calls and background processes run (greatly improves tail latency)

### CODE REVIEW

**Comment by ppenna** [comment]
> Use demi_wait_any instead of regular demi_wait to reduce the number of demi calls and background processes run (greatly improves tail latency)

@stolet I cannot find this enhancement in this PR.

**Comment by stolet** [comment]
For some reason the demi_wait enhancement didn't show up. I will create another pull request in a different branch so that we can add this change independently. I noticed that the other changes in this PR were broken up and added in different merges.

File: shim/src/utils.h
Comment (ppenna) [inline shim/src/utils.h:1]: Changes on this file are not required.

File: shim/src/queue.c
Comment (ppenna) [inline shim/src/queue.c:1]: Changes on this file are not required.

File: shim/src/interpose.c
Comment (ppenna) [inline shim/src/interpose.c:1]: Why are these changes required?

File: shim/src/glue.c
Comment (ppenna) [inline shim/src/glue.c:1]: Changes on this file are not required.

Comment (stolet) [reply shim/src/glue.c:1]: It seems like there are a few conflicting/duplicate changes after the merges of #1357, #1358 and #1359. I separated the remaining change in another PR #1373. Perhaps we can close this one and just merge #1373 to avoid confusion.
